[BEAM-5964] Add ClickHouseIO.Write

kanterov opened the pull request:
[ClickHouse](https://clickhouse.yandex) is open-source columnar DBMS for OLAP. It allows analysis of data that is updated in real time. The project was released as open-source software under the Apache 2 license in June 2016.

See [BEAM-5964](https://issues.apache.org/jira/browse/BEAM-5964) for more details.

Features that are missing (to be added to JIRA):

- [x] java doc with examples and polish public API
- [x] add example (and test) for writing POJO-based PCollections
- [ ] integration tests for writing into `Distributed` tables
- [ ] integration tests and examples for writing unbounded PCollections
- [ ] reading in `ClickHouseIO.Read`
- [ ] validate that input schema is compatible with table schema (see Cast PTransform)
- [ ] support for `Enum8` and `Enum16` data types
- [ ] support for `Decimal` data type
- [ ] support for `UUID` data type
- [ ] support for experimental `LowCardinality(String)` data type


------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

kanterov commented on sdks/java/io/clickhouse/OWNERS:
```diff
@@ -0,0 +1,4 @@
+# See the OWNERS docs at https://s.apache.org/beam-owners
+
+reviewers:
+  - kanterov
```
I put my name here, but I'm not a beam committer. How do we find the owner? @iemejia

kanterov commented:
@serebrserg @alexey-milovidov can you please take a look at ClickHouse-specific parts: integration with clickhouse-jdbc, atomic and idempotent inserts

serebrserg commented on sdks/java/io/clickhouse/src/main/java/org/apache/beam/sdk/io/clickhouse/ClickHouseIO.java:
```diff
@@ -0,0 +1,364 @@
+        }
+
+        // findbugs doesn't like it in try block
+        rs.close();
```
I think this should be in finally block.

kanterov replied:
Initially I did inner try-with-resources, but it doesn't work with findbugs for some reason.

serebrserg commented on sdks/java/io/clickhouse/src/main/java/org/apache/beam/sdk/io/clickhouse/ClickHouseIO.java:
```diff
@@ -0,0 +1,364 @@
+                              // bundle is finished and queue is drained, time to go
+                              break;
+                            } else {
+                              // timeout, let's try again
```
From my point of view this might be cleaner with take() and interruption on finishBundle, or sending a poison object through queue on finishBundle. Anyway, this is not a point why you asked for my comments, so this is just a suggestion.

Overall regarding queue, I don't see other options for streaming insert with current WriteFn interface now (what I see from annotations). ClickHouseRowBinaryStream had been stable for a long time, so I don't think it is going to disappear unexpectedly. 
The obvious alternative is to collect data in memory and then send the whole batch through jdbc's PreparedStatement.executeBatch() (this is cleaner from interface point of view, but needs memory and may be slower, at least now, since it currently uses tsv as serialization).
I'm not aware of Beam's typical setup and workload, so I'm not sure which option is best in this case.

serebrserg commented on sdks/java/io/clickhouse/src/main/java/org/apache/beam/sdk/io/clickhouse/ClickHouseWriter.java:
```diff
@@ -0,0 +1,134 @@
+      ClickHouseRowBinaryStream stream, ColumnType columnType, Object value) throws IOException {
+
+    if (value == null) {
+      stream.writeByte((byte) 1);
```
I suppose this is better to be implemented in clickhouse-jdbc. I think this is too much of implementation detail of ClickHouse to be anywhere rather than driver.

kanterov replied:
@serebrserg agree, submitted a pull request

chamikaramj commented:
Happy to take a look.

kanterov commented:
@chamikaramj did you have a chance to take a look?

chamikaramj commented:
Sorry about the delay. Will get to it in next few days.

chamikaramj commented on sdks/java/io/clickhouse/src/main/javacc/ColumnTypeParser.jj:
```diff
@@ -0,0 +1,181 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
```
Could you explain what this is needed ? (probably add a comment)

kanterov replied:
Added a comment to generated class Javadoc

kanterov commented on sdks/java/io/clickhouse/src/main/java/org/apache/beam/sdk/io/clickhouse/ClickHouseIO.java:
```diff
@@ -0,0 +1,364 @@
+                        while (true) {
+                          Row row;
+                          try {
+                            row = queue.poll(1, TimeUnit.SECONDS);
```
Sorry, forgot to include context that other connectors do retries and an inner batching, because, as I understand, there is guarantee that data will fit the memory.

With this approach we stream data as it arrives, so we don't have to keep it in memory.

kanterov commented on sdks/java/io/clickhouse/src/main/java/org/apache/beam/sdk/io/clickhouse/ClickHouseIO.java:
```diff
@@ -0,0 +1,364 @@
+    }
+
+    /**
+     * For INSERT queries in the replicated table, specifies that deduplication of inserting blocks
```
This setting is enabled by default, and it used only by ClickHouse, but not the connector. Probably we should completely remove it to avoid confusion.

kanterov commented on sdks/java/io/clickhouse/src/main/java/org/apache/beam/sdk/io/clickhouse/ClickHouseIO.java:
```diff
@@ -0,0 +1,364 @@
+    }
+
+    /**
+     * For INSERT queries in the replicated table, wait writing for the specified number of replicas
```
Default behavior depends on server settings, but the default server settings have this setting disabled. I think it worth mentioning :). 👍

kanterov commented on sdks/java/io/clickhouse/src/main/java/org/apache/beam/sdk/io/clickhouse/ClickHouseIO.java:
```diff
@@ -0,0 +1,364 @@
+    @FinishBundle
+    public void finishBundle() throws ExecutionException, InterruptedException {
+      bundleFinished.set(true);
+      insert.get(); // drain queue
```
Deduplication is done on the server side. When a block is inserted, replicas put checksum of the block to a ring buffer on top of ZooKeeper. In our case, one bundle is one block, so deduplication should work as soon as bundles don't change between retries, because even splitting a bundle will break deduplication. I see there is `@RequiresStableInput` but wasn't sure if it is needed in this case.

kanterov commented on sdks/java/io/clickhouse/src/main/java/org/apache/beam/sdk/io/clickhouse/ClickHouseWriter.java:
```diff
@@ -0,0 +1,134 @@
+import ru.yandex.clickhouse.util.ClickHouseRowBinaryStream;
+
+/** Writes Rows and field values using {@link ClickHouseRowBinaryStream}. */
+@Experimental(Experimental.Kind.SCHEMAS)
```
I did `SCHEMAS` because it uses `Row` and `Schema` APIs. Agree that `SOURCE_SINK` fits better.

chamikaramj commented:
Regarding fusion, assume there is step A before ParDo(WriteFn). Some runners (for example, Dataflow) may fuse these steps. So if step A generate a different set of data in a rerun (due to a failure) we may insert two different sets of data. The way to fix this is to break fusion. This can be done by adding a Reshuffle.viaRandomKey() step before writing.
https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Reshuffle.java#L64

It should be fine to do retries within a DoFn as long as this does not result in duplicate data in the data store.
In fact, this is encouraged, (for example, you might want to use exponential backoff in retries to prevent the data store from being overloaded). Bundle level retries (four times for batch) is a runner specific last resort effort and IO connectors should not depend on this behavior.

kanterov commented:
@chamikaramj I see, I'm a bit worried about shuffling data, because it would significantly slow down pipelines, I see a typical use case for this connector would be reading AVRO files, and uploading them to ClickHouse. As I understand, such input should be deterministic. Do you think it would make sense to make shuffling optional?

chamikaramj commented:
Actually, I take it back. I don't think we need a preceding shuffle here since retries do not write duplicate data for a deterministic input.

chamikaramj commented:
Please let me know when the current comments are addressed.

chamikaramj commented:
Gleb, any updates ?

Please let me know if you have any questions on my comments.

kanterov commented:
@chamikaramj sorry for the delay, I'm addressing comments, and it's close to being finished.

I went with a more straightforward approach we discussed in comments. It works well so far, and the code became much cleaner. I'm doing tests to see what is the impact on performance, and I will get back to you when it's more clear.

chamikaramj commented:
No worries. Thanks for the update.

kanterov commented:
Run Portable_Python PreCommit

kanterov commented:
@chamikaramj can you please take another look? 

I've addressed your comments. It seems Portable_Python and Website are failing for a reason not related to this PR.

chamikaramj commented:
Thanks. I'll take another look today.

chamikaramj commented:
LGTM. Thanks.

Do you hope to add an integration test as well (can be a separate PR) ?

Please fixup commits for merging.

chamikaramj commented:
Retest this please

kanterov commented:
@chamikaramj Thanks! I've rebased commits, and I can squash them if you don't mind.

What do you have in mind for integration tests? Now there are tests using direct-runner and ClickHouse running in Docker.

kanterov commented:
@chamikaramj I did rebase and split into 4 commits, before squashing I moved commits around to minimize unnecessary diffs between final commits. Each of 4 commits compiles, tests and style checks pass.

kanterov commented:
Run Java PreCommit

kanterov commented:
@chamikaramj is there anything left before we can merge?

chamikaramj commented:
Thanks. Merging.

kanterov commented:
@chamikaramj @serebrserg great job, thanks a lot for your feedback!

kanterov commented:
@kennknowles given this change is isolated, and PMC didn't vote for 2.10 can we cherry-pick it?

kennknowles commented:
@kanterov Yes, feel free to propose a cherrypick PR. This was clearly pretty much done when the release proposal thread started.
